internal/femto/femto.go

```diff
@@ -3,124 +3,145 @@ package femto
+
+func (femto *Femto) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	femto.mux.ServeHTTP(w, r)
+}
+
+func OnPost[T any](femto *Femto, pattern string, handle PostFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doPost(femto, writer, request, handle)
 	})
 }
 
-func OnGet[T any](f *Femto, pattern string, handle GetFunc[T]) {
-	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		doGet(f, w, r, handle)
+func OnGet[T any](femto *Femto, pattern string, handle GetFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doGet(femto, writer, request, handle)
 	})
 }
 
-func doGet[T any](f *Femto, w http.ResponseWriter, r *http.Request, handle GetFunc[T]) {
-	reqNo := f.nextReqNo()
-	log := f.logger().With(slog.Uint64("req_no", reqNo))
+func generateErrorLogger(log *slog.Logger, writer http.ResponseWriter) func(ctx string, status int, e error) {
+	return func(ctx string, status int, e error) {
+		log.Error(ctx, "err", e)
+		http.Error(writer, e.Error(), status)
+	}
+}
+
+func doGet[T any](femto *Femto, writer http.ResponseWriter, request *http.Request, handle GetFunc[T]) {
+	reqNo := femto.nextReqNo()
+	log := femto.logger().With(slog.Uint64("req_no", reqNo))
 
-	log.Info("New Request", "method", r.Method, "url", r.URL, "from", r.RemoteAddr)
+	log.Info("New Request", "method", request.Method, "url", request.URL, "from", request.RemoteAddr)
 
-	if r.Method != http.MethodGet {
+	if request.Method != http.MethodGet {
 		log.Info("Wanted GET, returned 405")
-		http.Error(w, "Expected GET", http.StatusMethodNotAllowed)
+		http.Error(writer, "Expected GET", http.StatusMethodNotAllowed)
 		return
 	}
 
-	ise := func(ctx string, e error) {
-		log.Error(ctx, "err", e)
-		http.Error(w, e.Error(), http.StatusInternalServerError)
+	data, err := handle(request, log)
+
+	if data == nil {
+		data = &HTTPResponseContent[T]{}
 	}
 
-	data, err := handle(r, log)
+	if data.Status == 0 {
+		data.Status = http.StatusInternalServerError
+	}
+
+	ise := generateErrorLogger(log, writer)
+
 	if err != nil {
-		// Handle authentication related errors
-		if errors.Is(err, NotAuthenticatedError) {
-			log.Info("Invalid attempt at authentication")
-			http.Error(w, "Not authenticated", http.StatusUnauthorized)
-			return
-		}
-		ise("application error", err)
+		ise("There was an error in the handler", data.Status, err)
 		return
 	}
 
-	jsonb, err := json.Marshal(data)
+	jsonb, err := json.Marshal(data.Body)
 	if err != nil {
-		ise("marshaling to json failed", err)
+		ise("There was an error in trying to parse the handler's response into JSON", data.Status, err)
 		return
 	}
 
-	_, err = w.Write(jsonb)
+	for key, value := range data.Headers {
+		writer.Header().Add(key, value)
+	}
+
+	_, err = writer.Write(jsonb)
+
 	if err != nil {
-		ise("writing failed", err)
+		ise("There was an error in trying to respond to the user", data.Status, err)
```

Comment:
This is less usefull, as it could now be any error. This should explicity say it was a problem writing.

Reply:
Responding to the user, and writing, is the same thing, no?

Reply:
In my mind, "responding to the user" could be anywhere in the entire handler. I'd like the logs to note that it's an IO error around writing the responce (and not for example calculating it)

Reply:
Sure, should we say an error in _communicating_ to the user then?

Reply:
why not just writing?

Reply:
Writing is server-specific (we write to a socket), not client-specific (the server is communicating with us) and this is a client-facing API.

Reply:
But if we're failing to write, the client isn't going to see it. These errors also go to our logs, and they need to be usefull their aswell.

(but whatever, we've burnt too many cycles on this, let's get the important things right)